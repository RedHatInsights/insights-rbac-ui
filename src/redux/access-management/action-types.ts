export const GET_PRINCIPAL_ACCESS = 'GET_PRINCIPAL_ACCESS' as const;
