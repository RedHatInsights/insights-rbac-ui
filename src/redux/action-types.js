export const FETCH_GROUP = 'FETCH_GROUP';
export const FETCH_SYSTEM_GROUP = 'FETCH_SYSTEM_GROUP';
export const FETCH_GROUPS = 'FETCH_GROUPS';

export const ADD_GROUP = 'ADD_GROUP';
export const UPDATE_GROUP = 'UPDATE_GROUP';
export const REMOVE_GROUPS = 'REMOVE_GROUPS';

export const FETCH_USERS = 'FETCH_USERS';

export const ADD_ROLE = 'ADD_ROLE';
export const FETCH_ROLE = 'FETCH_ROLE';
export const FETCH_ROLES = 'FETCH_ROLES';
export const REMOVE_ROLE = 'REMOVE_ROLE';
export const RESET_SELECTED_ROLE = 'RESET_SELECTED_ROLE';
export const FETCH_ROLES_FOR_WIZARD = 'FETCH_ROLES_FOR_WIZARD';

export const FETCH_GROUP_POLICIES = 'FETCH_GROUP_POLICIES';
export const FETCH_POLICY = 'FETCH_POLICY';
export const ADD_POLICY = 'ADD_POLICY';
export const UPDATE_POLICY = 'UPDATE_POLICY';
export const REMOVE_POLICY = 'REMOVE_POLICY';
export const RESET_SELECTED_GROUP = 'RESET_SELECTED_GROUP';

export const REMOVE_MEMBERS_FROM_GROUP = 'REMOVE_MEMBERS_FROM_GROUP';
export const ADD_MEMBERS_TO_GROUP = 'ADD_MEMBERS_TO_GROUP';

export const REMOVE_ROLES_FROM_GROUP = 'REMOVE_ROLES_FROM_GROUP';
export const ADD_ROLES_TO_GROUP = 'ADD_ROLES_TO_GROUP';
export const FETCH_ROLES_FOR_GROUP = 'FETCH_ROLES_FOR_GROUP';
export const FETCH_MEMBERS_FOR_GROUP = 'FETCH_MEMBERS_FOR_GROUP';
export const FETCH_ADD_ROLES_FOR_GROUP = 'FETCH_ADD_ROLES_FOR_GROUP';
export const FETCH_ROLE_FOR_USER = 'FETCH_ROLE_FOR_USER';
export const FETCH_ROLE_FOR_PRINCIPAL = 'FETCH_ROLE_FOR_PRINCIPAL';
export const UPDATE_ROLE = 'UPDATE_ROLE';

export const GET_PRINCIPAL_ACCESS = 'GET_PRINCIPAL_ACCESS';

export const LIST_PERMISSIONS = 'LIST_PERMISSIONS';
export const LIST_APPLICATION_OPTIONS = 'LIST_APPLICATION_OPTIONS';
export const LIST_RESOURCE_OPTIONS = 'LIST_RESOURCE_OPTIONS';
export const LIST_OPERATION_OPTIONS = 'LIST_OPERATION_OPTIONS';

export const GET_RESOURCE_DEFINITIONS = 'GET_RESOURCE_DEFINITIONS';
export const GET_RESOURCE = 'GET_RESOURCE';
