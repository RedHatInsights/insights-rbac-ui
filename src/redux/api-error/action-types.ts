export const API_ERROR = 'API_ERROR' as const;
