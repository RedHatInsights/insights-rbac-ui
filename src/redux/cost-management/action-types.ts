export const FETCH_RESOURCE_DEFINITIONS = 'FETCH_RESOURCE_DEFINITIONS' as const;
export const FETCH_RESOURCE = 'FETCH_RESOURCE' as const;
export const EXPAND_SPLATS = 'EXPAND_SPLATS' as const;
export const RESET_EXPAND_SPLATS = 'RESET_EXPAND_SPLATS' as const;
