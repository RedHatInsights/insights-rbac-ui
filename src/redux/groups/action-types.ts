// Group core actions
export const FETCH_GROUP = 'FETCH_GROUP' as const;
export const FETCH_ADMIN_GROUP = 'FETCH_ADMIN_GROUP' as const;
export const FETCH_SYSTEM_GROUP = 'FETCH_SYSTEM_GROUP' as const;
export const INVALIDATE_SYSTEM_GROUP = 'INVALIDATE_SYSTEM_GROUP' as const;
export const FETCH_GROUPS = 'FETCH_GROUPS' as const;

export const ADD_GROUP = 'ADD_GROUP' as const;
export const UPDATE_GROUP = 'UPDATE_GROUP' as const;
export const REMOVE_GROUPS = 'REMOVE_GROUPS' as const;
export const UPDATE_GROUPS_FILTERS = 'UPDATE_GROUPS_FILTERS' as const;
export const RESET_SELECTED_GROUP = 'RESET_SELECTED_GROUP' as const;

// Group members
export const ADD_MEMBERS_TO_GROUP = 'ADD_MEMBERS_TO_GROUP' as const;
export const REMOVE_MEMBERS_FROM_GROUP = 'REMOVE_MEMBERS_FROM_GROUP' as const;
export const FETCH_MEMBERS_FOR_GROUP = 'FETCH_MEMBERS_FOR_GROUP' as const;
export const FETCH_MEMBERS_FOR_EXPANDED_GROUP = 'FETCH_MEMBERS_FOR_EXPANDED_GROUP' as const;

// Group roles
export const ADD_ROLES_TO_GROUP = 'ADD_ROLES_TO_GROUP' as const;
export const REMOVE_ROLES_FROM_GROUP = 'REMOVE_ROLES_FROM_GROUP' as const;
export const FETCH_ROLES_FOR_GROUP = 'FETCH_ROLES_FOR_GROUP' as const;
export const FETCH_ROLES_FOR_EXPANDED_GROUP = 'FETCH_ROLES_FOR_EXPANDED_GROUP' as const;
export const FETCH_ADD_ROLES_FOR_GROUP = 'FETCH_ADD_ROLES_FOR_GROUP' as const;

// Group service accounts
export const ADD_SERVICE_ACCOUNTS_TO_GROUP = 'ADD_SERVICE_ACCOUNTS_TO_GROUP' as const;
export const REMOVE_SERVICE_ACCOUNTS_FROM_GROUP = 'REMOVE_SERVICE_ACCOUNTS_FROM_GROUP' as const;
export const FETCH_SERVICE_ACCOUNTS_FOR_GROUP = 'FETCH_SERVICE_ACCOUNTS_FOR_GROUP' as const;
