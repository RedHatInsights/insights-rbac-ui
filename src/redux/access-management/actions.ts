import { GET_PRINCIPAL_ACCESS } from './action-types';
import { GetPrincipalAccessParams, getPrincipalAccess as getPrincipalAccessHelper } from './helper';

// Using global ReduxAction interface from store.d.ts
export const getPrincipalAccess = (apiProps: GetPrincipalAccessParams): ReduxAction<Promise<unknown>> => ({
  type: GET_PRINCIPAL_ACCESS,
  payload: getPrincipalAccessHelper(apiProps),
});
