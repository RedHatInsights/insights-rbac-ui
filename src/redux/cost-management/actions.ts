import { FETCH_RESOURCE, FETCH_RESOURCE_DEFINITIONS } from './action-types';
import { getResource, getResourceDefinitions } from './helper';

// Based on cost API usage patterns, apiProps is typically a configuration object
interface ApiProps {
  [key: string]: unknown;
}

// Using global ReduxAction interface from store.d.ts
export const fetchResourceDefinitions = (apiProps: ApiProps): ReduxAction<Promise<unknown>> => ({
  type: FETCH_RESOURCE_DEFINITIONS,
  payload: getResourceDefinitions(apiProps),
});

export const fetchResource = (apiProps: ApiProps): ReduxAction<Promise<unknown>> => ({
  type: FETCH_RESOURCE,
  payload: getResource(apiProps),
});
