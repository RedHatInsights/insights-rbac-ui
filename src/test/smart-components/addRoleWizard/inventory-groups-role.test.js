import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { BrowserRouter } from 'react-router-dom';
import InventoryGroupsRole from '../../../features/roles/add-role/inventory-groups-role';
import { testInventoryGroups } from './inventory-groups-test-data';
import * as InventoryActions from '../../../redux/inventory/actions';
import { FETCH_INVENTORY_GROUPS } from '../../../redux/inventory/action-types';

const initialStateWithPermissions = {
  inventoryReducer: {
    isLoading: false,
    resourceTypes: {
      'inventory:hosts:read': testInventoryGroups.reduce((acc, curr) => ({ ...acc, [curr.name]: { ...curr } }), {}),
      'inventory:hosts:write': testInventoryGroups.reduce((acc, curr) => ({ ...acc, [curr.name]: { ...curr } }), {}),
    },
    total: testInventoryGroups.length,
  },
};

jest.mock('@data-driven-forms/react-form-renderer/use-form-api', () => {
  return {
    __esModule: true,
    default: () => ({
      getState: jest.fn().mockReturnValue({
        values: {
          'add-permissions-table': [
            {
              uuid: 'inventory:hosts:read',
              requires: [],
            },
            {
              uuid: 'inventory:groups:write',
              requires: [],
            },
          ],
          'role-description': 'Be able to read and edit Account Staleness and Culling data.',
          'role-name': 'Account Staleness and Culling Administrator',
          'role-type': 'create',
          'role-uuid': '218ed86a-08f9-4d5d-aeac-7539eed882cd',
        },
      }),
      change: jest.fn(),
    }),
  };
});

jest.mock('@data-driven-forms/react-form-renderer/use-field-api', () => {
  return {
    __esModule: true,
    default: () => ({
      input: {
        value: ['inventory:hosts:read', 'inventory:groups:write'],
        name: 'inventory-groups-role',
        onChange: jest.fn(),
        onBlur: jest.fn(),
        onFocus: jest.fn(),
      },
    }),
  };
});

jest.mock('@unleash/proxy-client-react', () => ({
  __esModule: true,
  useFlag: () => true,
}));

const renderComponent = (store) => {
  return render(
    <React.Fragment>
      <Provider store={store}>
        <BrowserRouter>
          <InventoryGroupsRole />
        </BrowserRouter>
      </Provider>
    </React.Fragment>,
  );
};

describe('Inventory groups role', () => {
  let mockStore;

  const fetchInventoryGroupsSpy = jest.spyOn(InventoryActions, 'fetchInventoryGroups');

  beforeEach(() => {
    mockStore = configureStore();
  });

  afterEach(() => {
    fetchInventoryGroupsSpy.mockReset();
  });

  test('Add permissions to group renders without failing', async () => {
    fetchInventoryGroupsSpy.mockImplementationOnce(() => ({
      type: FETCH_INVENTORY_GROUPS,
      payload: Promise.resolve({}),
    }));
    const store = mockStore(initialStateWithPermissions);

    const renderedResults = renderComponent(store);

    expect(renderedResults.getByText('Permissions')).toBeInTheDocument();
    expect(renderedResults.asFragment()).toMatchSnapshot();
  });

  test('Display available permission groups successfully', async () => {
    fetchInventoryGroupsSpy.mockImplementationOnce(() => ({
      type: FETCH_INVENTORY_GROUPS,
      payload: Promise.resolve({}),
    }));

    const store = mockStore(initialStateWithPermissions);
    renderComponent(store);

    fireEvent.click(screen.getAllByLabelText('Select workspaces to add permissions for')[0]);

    expect(screen.getByText('fooBar')).toBeInTheDocument();
  });

  test('Selecting group for permission and Copy to all to other permissions successfully', async () => {
    fetchInventoryGroupsSpy.mockImplementationOnce(() => ({
      type: FETCH_INVENTORY_GROUPS,
      payload: Promise.resolve({}),
    }));

    const store = mockStore(initialStateWithPermissions);
    const renderedResults = renderComponent(store);

    fireEvent.click(screen.getAllByLabelText('Select workspaces to add permissions for')[0]);
    expect(screen.getByText('fooBar')).toBeInTheDocument();
    fireEvent.click(screen.getByText('fooBar'));

    const badgesAfterSelection = screen.getAllByText('1', { selector: '.pf-v5-c-badge' });
    expect(badgesAfterSelection).toHaveLength(1);

    // Opening the second groups selection
    fireEvent.click(renderedResults.getByText('Copy to all'));
    expect(screen.getAllByLabelText('Select workspaces to add permissions for')[1]);
    fireEvent.click(screen.getAllByLabelText('Select workspaces to add permissions for')[1]);

    const badgesAfterCopyAll = screen.getAllByText('1', { selector: '.pf-v5-c-badge' });
    expect(badgesAfterCopyAll).toHaveLength(2);
  });

  test('Cancelling groups selection clears selections successfully', async () => {
    fetchInventoryGroupsSpy.mockImplementationOnce(() => ({
      type: FETCH_INVENTORY_GROUPS,
      payload: Promise.resolve({}),
    }));

    const store = mockStore(initialStateWithPermissions);
    const renderedResults = renderComponent(store);

    fireEvent.click(screen.getAllByLabelText('Select workspaces to add permissions for')[0]);
    expect(renderedResults.getByText('fooBar')).toBeInTheDocument();
    fireEvent.click(screen.getByText('fooBar'));

    const badgesBeforeClear = screen.getAllByText('1', { selector: '.pf-v5-c-badge' });
    expect(badgesBeforeClear).toHaveLength(1);

    await waitFor(() => {
      const clearButton = screen.getByLabelText('Clear all');
      expect(clearButton).toBeInTheDocument();
    });

    fireEvent.click(screen.getByLabelText('Clear all'));

    const badgesAfterClear = screen.queryAllByLabelText('Clear all');
    expect(badgesAfterClear).toHaveLength(0);
  });
});
