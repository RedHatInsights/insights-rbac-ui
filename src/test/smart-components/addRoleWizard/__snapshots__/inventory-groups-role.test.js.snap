// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inventory groups role Add permissions to group renders without failing 1`] = `
<DocumentFragment>
  <div
    class="pf-v5-l-grid pf-m-gutter"
  >
    <div
      class="pf-v5-l-grid__item pf-m-3-col-on-md rbac-m-hide-on-sm"
    >
      <h4
        class="rbac-bold-text pf-v5-u-mt-sm"
        data-ouia-component-id="OUIA-Generated-Text-1"
        data-ouia-component-type="PF5/Text"
        data-ouia-safe="true"
        data-pf-content="true"
      >
        Permissions
      </h4>
    </div>
    <div
      class="pf-v5-l-grid__item pf-m-6-col-on-md pf-m-9-col-on-lg rbac-m-hide-on-sm"
    >
      <h4
        class="rbac-bold-text pf-v5-u-mt-sm"
        data-ouia-component-id="OUIA-Generated-Text-2"
        data-ouia-component-type="PF5/Text"
        data-ouia-safe="true"
        data-pf-content="true"
      >
        Workspaces definition
      </h4>
    </div>
    <div
      class="pf-v5-l-grid"
    >
      <div
        class="pf-v5-l-grid__item pf-m-3-col-on-md"
      >
        <div
          class="pf-v5-c-form__group"
        >
          <div
            class="pf-v5-c-form__group-label"
          >
            <label
              class="pf-v5-c-form__label"
            >
              <span
                class="pf-v5-c-form__label-text"
              >
                hosts:read
              </span>
              <span
                aria-hidden="true"
                class="pf-v5-c-form__label-required"
              >
                 *
              </span>
            </label>
             
          </div>
          <div
            class="pf-v5-c-form__group-control"
          />
        </div>
      </div>
      <div
        class="pf-v5-l-grid__item pf-m-7-col-on-md"
      >
        <div
          style="display: contents;"
        >
          <div
            class="pf-v5-c-menu-toggle pf-m-full-width pf-m-typeahead"
          >
            <div
              class="pf-v5-c-text-input-group pf-m-plain"
            >
              <div
                autocomplete="off"
                class="pf-v5-c-text-input-group__main"
              >
                <span
                  class="pf-v5-c-text-input-group__text"
                >
                  <input
                    aria-expanded="false"
                    aria-label="Type to filter"
                    class="pf-v5-c-text-input-group__text-input"
                    placeholder="Select workspaces"
                    role="combobox"
                    type="text"
                    value=""
                  />
                </span>
              </div>
              <div
                class="pf-v5-c-text-input-group__utilities"
              />
            </div>
            <button
              aria-expanded="false"
              aria-label="Select workspaces to add permissions for"
              class="pf-v5-c-menu-toggle__button"
              tabindex="-1"
              type="button"
            >
              <span
                class="pf-v5-c-menu-toggle__controls"
              >
                <span
                  class="pf-v5-c-menu-toggle__toggle-icon"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 320 512"
                    width="1em"
                  >
                    <path
                      d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                    />
                  </svg>
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="pf-v5-l-grid__item pf-m-2-col-on-md"
      >
        <button
          aria-disabled="false"
          class="pf-v5-c-button pf-m-link"
          data-ouia-component-id="OUIA-Generated-Button-link-1"
          data-ouia-component-type="PF5/Button"
          data-ouia-safe="true"
          type="button"
        >
          Copy to all
        </button>
      </div>
    </div>
    <div
      class="pf-v5-l-grid"
    >
      <div
        class="pf-v5-l-grid__item pf-m-3-col-on-md"
      >
        <div
          class="pf-v5-c-form__group"
        >
          <div
            class="pf-v5-c-form__group-label"
          >
            <label
              class="pf-v5-c-form__label"
            >
              <span
                class="pf-v5-c-form__label-text"
              >
                groups:write
              </span>
              <span
                aria-hidden="true"
                class="pf-v5-c-form__label-required"
              >
                 *
              </span>
            </label>
             
          </div>
          <div
            class="pf-v5-c-form__group-control"
          />
        </div>
      </div>
      <div
        class="pf-v5-l-grid__item pf-m-7-col-on-md"
      >
        <div
          style="display: contents;"
        >
          <div
            class="pf-v5-c-menu-toggle pf-m-full-width pf-m-typeahead"
          >
            <div
              class="pf-v5-c-text-input-group pf-m-plain"
            >
              <div
                autocomplete="off"
                class="pf-v5-c-text-input-group__main"
              >
                <span
                  class="pf-v5-c-text-input-group__text"
                >
                  <input
                    aria-expanded="false"
                    aria-label="Type to filter"
                    class="pf-v5-c-text-input-group__text-input"
                    placeholder="Select workspaces"
                    role="combobox"
                    type="text"
                    value=""
                  />
                </span>
              </div>
              <div
                class="pf-v5-c-text-input-group__utilities"
              />
            </div>
            <button
              aria-expanded="false"
              aria-label="Select workspaces to add permissions for"
              class="pf-v5-c-menu-toggle__button"
              tabindex="-1"
              type="button"
            >
              <span
                class="pf-v5-c-menu-toggle__controls"
              >
                <span
                  class="pf-v5-c-menu-toggle__toggle-icon"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 320 512"
                    width="1em"
                  >
                    <path
                      d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                    />
                  </svg>
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="pf-v5-l-grid__item pf-m-2-col-on-md"
      />
    </div>
  </div>
</DocumentFragment>
`;
